File: Day 27/solution.kt
class RandomizedSet() {

    /** Initialize your data structure here. */
    val map = HashMap<Int, Int>() // key: `val`, value:  index
    val arr = ArrayList<Int>()
    val random = Random()

    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    fun insert(`val`: Int): Boolean {
        if (map.contains(`val`)) return false

        arr.add(`val`)
        map[`val`] = arr.size - 1

        return true
    }

    /** Removes a value from the set. Returns true if the set contained the specified element. */
    fun remove(`val`: Int): Boolean {
        if (!map.contains(`val`)) return false

        val index = map[`val`]!!
        val last = arr.last()

        map[last] = index
        map.remove(`val`)

        arr[index] = last
        arr.removeAt(arr.size - 1)
        
        return true
    }

    /** Get a random element from the set. */
    fun getRandom(): Int {
        val index = random.nextInt(arr.size)

        return arr[index]
    }
}

// C++

class RandomizedSet {
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if (map.count(val)) return false;
        arr.push_back(val);
        map[val] = arr.size() - 1;
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if (!map.count(val)) return false;
        int index = map[val];
        int last = arr.back();
        arr[index] = last;
        arr.pop_back();
        map[last] = index;
        map.erase(val);
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        int index = rand() % arr.size();
        return arr[index];
    }
private:
    unordered_map<int, int> map;
    vector<int> arr;
};

File: Day 27/solution.c
class RandomizedSet {
private:
    unordered_map<int,int>mp;
    vector<int>nums;
public:
    /** Initialize your data structure here. */
    RandomizedSet() {
        //we already initialize
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if(mp.find(val) != mp.end())     // already present
            return false;
        nums.emplace_back(val); // push_back value;
        mp[val] = nums.size() - 1; //nums last index
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if(mp.find(val) == mp.end())    //not in the map
            return false;
        //The below 3 line to make complexity O(1)
        //vector::erase will take O(log n), so we shift our last element to other location and we remove the last element that will take O(1)
        int last_value = nums.back(); // Last value
        mp[last_value] = mp[val]; // reassign the new index
        nums[mp[val]] = last_value; //change the last value to other location
        
        nums.pop_back(); //It'll take O(1)
        mp.erase(val);
        return true;
    }
    
    /** Get a random element from the set. */
    int getRandom() {
        return nums[rand() % nums.size()]; //randomly generate index 
    }
};